### PR TITLE
crunch controller refactor

### PR DESCRIPTION
**Pull request by Nauna**
Now we can use the crunch controller to build the sections, wizardlets and closing logic outside of a wizard.

This will be necessary for: https://nanopay.atlassian.net/browse/NP-1481

### The calls to use will be something like:
var objWithCapsAndWizardlet = this.crunchController.getCapsAndWizardlets(capabilityId);
var usableForViewSections = this.crunchController.generateSections(objWithCapsAndWizardlet.wizCaps);
// to save 
-1 this.crunchController.finalOnClose(x, objWithCapsAndWizardlet.caps)
-2 forEach(objWithCapsAndWizardlet.wizSec, function(wizardlet) { wizardlet.save(); }

### CODE REVIEW

File: src/foam/u2/crunch/CrunchController.js [hunk 1]
```diff
@@ -88,89 +88,110 @@ foam.CLASS({
...
-            }).updateUCJ()
-          ))
-
+    function generateSections(generatedWizardlets) {
+      // return Promise.resolve(
+      return Promise.all([
+          generatedWizardlets.map((wizardlet) => {
+            this.AbstractSectionedDetailView.create({
+              of: wizardlet.of,
+            }).sections;
```
Comment (KernelDeimos) [inline src/foam/u2/crunch/CrunchController.js]: Does this value really get returned?

Comment (Nauna) [reply src/foam/u2/crunch/CrunchController.js]: good catch - fixed

**Review by KernelDeimos — APPROVED** [review]

File: src/foam/u2/crunch/CrunchController.js [hunk 2]
```diff
@@ -88,89 +88,105 @@ foam.CLASS({
...
+            .filter(cap => !! cap.of )
+            .map(cap =>
+              this.CapabilityWizardlet.create({ capability: cap }).updateUCJ())
             )
-          );
-          ctrl.add(this.Popup.create({ closeable: false }).tag({
+          ]).then((capAndSections) => {
+            return {
+              caps: capAndSections[0],
+              wizSec: capAndSections[1]
```
Comment (KernelDeimos) [inline src/foam/u2/crunch/CrunchController.js]: Inconsistent, would prefer `wizCaps, wizWizs` or `capabilities, wizardlets`

Comment (Nauna) [reply src/foam/u2/crunch/CrunchController.js]: changed to wizCaps

**Comment by Nauna** [comment]
just another note on usage on a view:
given usableForViewSections and objWithCapsAndWizardlet (defined above)

// for each usableForViewSections[indexI] ... gives you list of sections per capability
// then usableForViewSections[indexI][indexJ] is one section
{
        class: 'foam.u2.detail.SectionView',
        section: usableForViewSections[indexI][indexJ],
        data$: objWithCapsAndWizardlet.wizCaps[indexI].data$,
      }

ex of using crunchController in view look at foam.u2.wizard.StepWizardletController

File: src/foam/u2/crunch/CrunchController.js [hunk 3]
```diff
@@ -88,89 +88,105 @@ foam.CLASS({
...
-          );
-          ctrl.add(this.Popup.create({ closeable: false }).tag({
+          ]).then((capAndSections) => {
+            return {
+              caps: capAndSections[0],
+              wizCaps: capAndSections[1]
+                .filter(wizardSection =>
+                  wizardSection.ucj === null ||
+                  (
+                    ! foam.util.equals(wizardSection.ucj.status,
```
Comment (kgrgreer) [inline src/foam/u2/crunch/CrunchController.js]: Why are you using 'equals'? Are these values both stati? Couldn't you use use ===?

File: src/foam/u2/crunch/CrunchController.js [hunk 4]
```diff
@@ -88,89 +88,105 @@ foam.CLASS({
...
+                )
+            };
+          });
+      });
+    },
+
+    function generateSections(generatedWizardlets) {
+      return generatedWizardlets.map(wizardlet =>
+        this.AbstractSectionedDetailView.create({
+          of: wizardlet.of,
```
Comment (kgrgreer) [inline src/foam/u2/crunch/CrunchController.js]: no trailing commas

File: src/foam/u2/crunch/CrunchController.js (hunk 3 above)
Comment (Nauna) [reply src/foam/u2/crunch/CrunchController.js]: yikes sorry for that :(

File: src/foam/u2/crunch/CrunchController.js [hunk 5]
```diff
@@ -88,89 +88,103 @@ foam.CLASS({
...
+    },
 
+    async function launchWizard(capabilityId) {
+      var ucj = await this.userCapabilityJunctionDAO.find(
+        this.AND(
+          this.EQ(this.UserCapabilityJunction.SOURCE_ID, this.subject.user.id),
+          this.EQ(this.UserCapabilityJunction.TARGET_ID, capabilityId)
+        ));
+      if ( ucj ) {
+        var statusGranted = foam.util.equals(ucj.status, this.CapabilityJunctionStatus.GRANTED);
```
Comment (Nauna) [inline src/foam/u2/crunch/CrunchController.js]:
```suggestion
        var statusGranted = ucj.status === this.CapabilityJunctionStatus.GRANTED;
```

File: src/foam/u2/crunch/CrunchController.js [hunk 6]
```diff
@@ -88,89 +88,103 @@ foam.CLASS({
...
+        var statusPending = foam.util.equals(ucj.status, this.CapabilityJunctionStatus.PENDING);
```
Comment (Nauna) [inline src/foam/u2/crunch/CrunchController.js:173]:
```suggestion
        var statusPending = ucj.status === this.CapabilityJunctionStatus.PENDING;
```

**Pull request merged**